File: src/main/java/whilestd/Node.java
package whilestd;

import static whilestd.Utils.Nil;

public class Node extends BinTree {

    private BinTree left = null;
    private BinTree right = null;

    public Node() {}

    public Node(BinTree left, BinTree right) {
        this.left = left.clone();
        this.right = right.clone();
    }

    @Override
    protected int toInt() {
        return this.right != null ? 1 + this.right.toInt() : 0;
    }

    @Override
    protected boolean toBool() {
        return this.right != null;
    }

    @Override
    public String toString() {
        return this.left.toString() + this.right.toString();
    }

    @Override
    protected BinTree hd() {
        return this.left != null ? this.left.clone() : Nil();
    }

    @Override
    protected BinTree tl() {
        return this.right != null ? this.right.clone() : Nil();
    }


    @Override
    public BinTree clone() {
        return new Node(left.clone(), right.clone());
    }


    @Override
    public boolean equals(Object other) {
        if (other instanceof Node) {
            Node otherNode = (Node) other;
            return this.left.equals(otherNode.left) && this.right.equals(otherNode.right);
        }
        return false;
    }

    @Override
    protected void prettyPrint() {
        if ( this.left == null && this.right == null ) {
            System.out.print("nil");
        } else if (this.left.toString().equals("int")) {
            System.out.print(this.right.toInt());
        } else if (this.left.toString().equals("bool")) {
            System.out.print(this.right.toBool());
        } else if (this.left.toString().equals("string")) {
            System.out.print(this.right.toString());
        } else {
            System.out.print("(cons ");
            this.left.prettyPrint();
            System.out.print(" ");
            this.right.prettyPrint();
            System.out.print(")");
        }
    }
}

File: whilestd/src/Node.cpp
#include "whilestd/Node.h"
#include "whilestd/Leaf.h"
#include <memory>
#include <string>
#include <iostream>

namespace whilestd {

Node::Node(BinTreePtr left, BinTreePtr right)
    : left(std::move(left)), right(std::move(right)) {}

Node::operator int() const {
    if(isNil()) {
        return 0;
    }
    return 1 + static_cast<int>(*right);
}

Node::operator bool() const {
    return !isNil();
}

Node::operator std::string() const {
    if(isNil()) {
        return "";
    }
    return static_cast<std::string>(*left) + static_cast<std::string>(*right);
}

BinTreePtr Node::hd() const {
    if(isNil()) {
        return std::make_unique<Node>();
    }
    return std::move(left->clone());
}

BinTreePtr Node::tl() const {
    if(isNil()) {
        return std::make_unique<Node>();
    }
    return std::move(right->clone());
}

BinTreePtr Node::clone() const {
    if(isNil()) {
        return std::make_unique<Node>();
    }
    return std::make_unique<Node>(std::move(left->clone()), std::move(right->clone()));
}

bool Node::equals(const BinTreePtr& other) const {
    if (auto otherNode = dynamic_cast<Node*>(other.get())) {
        if (isNil()) {
            return otherNode->isNil();
        }
        return !otherNode->isNil()
            && left == otherNode->left
            && right == otherNode->right;
    }
    return false;
}

std::ostream& Node::pp(std::ostream& os) const {
    bool printed = false;
    if(isNil()) {
        os << "nil";
        printed = true;
    }
    else if (auto leftLeaf = dynamic_cast<Leaf*>(left.get())) {
        if (static_cast<std::string>(*leftLeaf) == "int") {
            return os << static_cast<int>(*right);
            printed = true;
        }
        else if (static_cast<std::string>(*leftLeaf) == "bool") {
            return os << std::boolalpha << static_cast<bool>(*right);
            printed = true;
        }
        else if (static_cast<std::string>(*leftLeaf) == "string") {
            return os << static_cast<std::string>(*right);
            printed = true;
        }
    }
    if (!printed) {
        os << "(cons " << left << " "  << right << ")";
    }
    return os;
}

bool Node::isNil() const {
    return left == nullptr && right == nullptr;
}

}
